File: Leetcode/动态规划/最大最小博弈/stone Game2.cpp
#include <iostream>
#include <cstdio>
using namespace std;

class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        int n = piles.size();
        vector<int> sum(n+1,0);
        for(int i = n-1;i>=0;i--)
            sum[i] = piles[i] + sum[i+1];
        vector<vector<int>> f(n+1,vector<int>(n+1));
        
        for(int i =n-1;i>=0;i--){
            for(int j =1;j<=n;j++){
                for(int k =1;k<=2*j && i+k<=n;k++){
                    f[i][j] = max(f[i][j], sum[i] - f[i+k][max(j,k)]);
                }
            }
        }
        return f[0][1];
    }
};

File: Leetcode/动态规划/最大最小博弈/stone Game2.py
class Solution(object):
    def stoneGameII(self, piles):
        """
        :type piles: List[int]
        :rtype: int
        """
        n = len(piles)
        d = {}
        sums = [0] * (n+1)
        for i in range(n-1,-1,-1):
            # print(piles[i])
            sums[i] = sums[i+1] + piles[i]
        
        # print(sums)
        
        def f(i,j):
            if i >= n:
                return 0
            if str(i)+"_"+str(j) in d:
                return d[str(i)+"_"+str(j)]
            max_value = 0
            
            for k in range(1,j*2+1):
                if k<=2*j and i+k<=n:
                    max_value = max(max_value, sums[i]-f(i+k,max(j,k)) )
            d[str(i)+"_"+str(j)] = max_value
            return d[str(i)+"_"+str(j)]
        return f(0,1)
        